### PR TITLE
Arbitrary buffer thickness

### PR DESCRIPTION
**Pull request by Niceno**
Dear Egor,

The new communication already proves to be rather robust.  In the last couple of hours I implemented communication buffer layers with arbitrary thickness.  What we had before was one layer, but now we can have two, three and even more, although I can hardly think of a reason why would one need more than two layers.

I would keep this feature as a _hidden_.  That means code will work exactly the same as before after this pull request.  However, if one changes the second parameter to the call to `Save_Subdomains` to a value bigger than one, more buffer layers will get created.

I would leave this as a hidden feature.  In the mid-run, it would be good to have two communication patterns coexisting, one which needs only one layer of cells (like about everything we did up to now) and the other one with more layers for beyond-the-Navier-Stokes models.  In other words, if one needs three layers of buffer cells for VOF, that should not force the remaining of the code to exchange two layers of cells for each synchronization.   I could be wrong, but the way I see it implemented in the mid-run would be:
```
    type(Buffer_Type) :: cell_send_1  ! send buffer with one layer of buffer cells
    type(Buffer_Type) :: cell_recv_1  ! receive buffer with one layer of buffer cells
    type(Buffer_Type) :: cell_send_2  ! send buffer with two layers of buffer cells
    type(Buffer_Type) :: cell_recv_2  ! receive buffer with two layers of buffer cells
```
This might also need two `MPI_COMM_WORLD`, but I am not sure at this moment.

Long story short: this pull request will not change anything in the functionality of the code ... yet one needing more than one layer of buffer cells can invoke this feature through hard coding and explore whatever he or she wants to explore.

    Cheers,

    Bojan

### CODE REVIEW

**Comment by Niceno** [comment]
Dear Egor, I fixed the bug with MPICH and backup files in parallel.  I gave a lengthy explanation above.

**Comment by Niceno** [comment]
Hi Egor, I forgot to push one file ... with this commit I also make sure that boundary cells on buffers are first in the list, so the fix in `Form_Maps` always work.

**Comment by Niceno** [comment]
Dear both,


I went to check what is going on, and have soon realized I forgot to merge
the last pull request on new_communication and arbitraary_buffer_thickness.

With this merged, I can't reproduce the bug Muhamed is talking about.

I suggest Muhamed pulls the latest sources and checks again.


    Cheers,


    Bojan
